relay-server/src/lib.rs

```diff
@@ -301,9 +303,30 @@ pub fn run(config: Config) -> anyhow::Result<()> {
     // information on all services.
     main_runtime.block_on(async {
         Controller::start(config.shutdown_timeout());
-        let service = ServiceState::start(config.clone())?;
+        let (service, mut join_handles) = ServiceState::start(config.clone())?;
         HttpServer::new(config, service.clone())?.start();
-        Controller::shutdown_handle().finished().await;
+
+        loop {
+            select! {
+                Some(res) = join_handles.next() => {
+                    match res {
+                        Ok(()) => {
+                            relay_log::trace!("Service exited normally.");
+                        }
+                        Err(e) => {
+                            if e.is_panic() {
+                                std::panic::resume_unwind(e.into_panic());
```

Comment:
Do we maybe want in a future iteration to define a respawn behavior of services? It might be tricky to make sure existing channels are re-setup.

Reply:
This actually re-triggers the panic and makes the process terminate. Respawning services is another option I would like to discuss on Monday, but it has its drawbacks (what if the service keeps panicking on every re-spawn?).

Reply:
Yes this is something I thought of, I feel like for that we should have some global retry counters or heuristics to know when it's not possible to restart a service anymore.